DS-195 remove the use of keystore apis on add a validator doc (#827)
ava-labs/avalanche-docs: docs/nodes/validate/add-a-validator.md

## Requirements

> You've completed [Run an Avalanche Node](../build/run-avalanche-node-manually.md) and are familiar with [Avalanche's architecture](../../overview/getting-started/avalanche-platform.md).In this tutorial, we use [AvalancheJS](../../apis/avalanchejs/README.md) and [Avalanche’s Postman collection](https://github.com/ava-labs/avalanche-postman-collection) to help us make API calls.

Reviewer: nit: you are missing a space after period:
```
md).In this tutorial, we use
```

#### Stake Amount

The sample code assigns the variable, `stakeAmount` to be the minimum amount of AVAX that a node must stake to become a validator(1 AVAX).

```ts
const stakeAmount: any = await pchain.getMinStake();
```

To stake a different amount, please replace the code above with a stake amount of 1 AVAX or greater. Below sets a new value of 10,000 AVAX (`10000000000000` GWEI). Value is set in GWEI format where `1,000,000,000` GWEI = 1 AVAX.

Please follow the example below for steps to change the stake amount:

Locate this line in the file

```js
  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
    ...
    stakeAmount.minValidatorStake,
    ...
  )
```

and change the default value to a new `BN` value, for example:

```js
  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
    ...
    new BN(1e12),
    ...
  )
```

#### Addresses

`toAddresses`

> An array of addresses as [Buffer](https://github.com/feross/buffer) who receive the staked tokens at the end of the staking period.

Reviewer 2: Please provide example of using [Buffer](https://github.com/feross/buffer) if that is the case.

Author: I realize this is confusing. I believe that it would be better to just remove the mention of `Buffer` as it pertains more to how data is parsed and doesnt have an effect on how addresses are passed in.

Locate this part of the code

```ts
let privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`
pKeychain.importKey(privKey)
```

and replace `privKey` with [private keys that you control](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/createKeypair.ts).

```ts
let privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`
pKeychain.importKey(privKey)
privKey = "PrivateKey-FLKnwy3eFr75LHw6mPn8RAsAcWPq4WD8erMWiTK4KsLvkuP2t"
pKeychain.importKey(privKey)
privKey = "PrivateKey-2AUhohZSb86GKXE3mTBNVQGDWYpgR3FGHh7r6gqFoUzXDBNCcA"
pKeychain.importKey(privKey)
priKey = "PrivateKey-JeiGENXTFw5NWzBR18iacf1mL18VosfHf5eURhvmqSiAZKLaC"
pKeychain.importKey(privKey)

const pAddressStrings: string[] = pchain.keyChain().getAddressStrings()
```

This example would create a keychain with 4 addresses:

```ts
[
  "P-custom18jma8ppw3nhx5r4ap8clazz0dps7rv5u9xde7p", // pAddressStrings[0]
  "P-custom186e8ee4qnhnazrnyv2k9u44607ux70syngdz6l", // pAddressStrings[1]
  "P-custom1pdu8d8ss5p329l4vtn23p25cjr57qt5yw2cnsn", // pAddressStrings[2]

>   "P-custom152s2upkznxeucwnp68wmg6wmy863qd2sz6ktww", // pAddressStrings[3]

Reviewer 2: This is not the correct address for Fuji as this section is still under the Fuji Workflow. @Julian-dev28 please change the text how to generate the correct address value for Fuji.

:::tip
You can create your own unix timestamp [here](https://www.unixtimestamp.com/) or by using the `UnixNow()` method
:::

To create your own start times, please follow the steps below:

Locate this line in the file

```ts
const startTime: BN = UnixNow().add(new BN(60 * 1));
const endTime: BN = startTime.add(new BN(26300000));
```

Change `startTime` and `endTime` to new `BN` values, for example:

```ts
const startTime: BN = new BN(1654656829); // Wed Jun 08 2022 02:53:49 GMT+0000
const endTime: BN = new BN(1662602029); // Thu Sep 08 2022 01:53:49 GMT+0000
```

#### Delegation Fee Rate

Avalanche allows for delegation of stake. This parameter is the percent fee this validator charges when others delegate stake to them. For example, if `delegationFeeRate` is `10` and someone delegates to this validator, then when the delegation period is over, 10% of the reward goes to the validator and the rest goes to the delegator, if this node meets the validation reward requirements.

#### Stake Amount

> Set the the proper staking amount when calling `pchain.buildAddValidatorTx` by replacing `stakeAmount.minValidatorStake` with a number such as `BN(1e12)` which is 10,000 AVAX, in the example code.

Reviewer 2: ```suggestion
Set the the proper staking amount in calling `pchain.buildAddValidatorTx` by replacing `stakeAmount.minValidatorStake` with a number in the unit of gwei, for example, `BN(1e12)` which is 10,000 AVAX.
```

You should a success message, and your balance should be updated.

Calling [`platform.getPendingValidators`](../../apis/avalanchego/apis/p-chain.md#platformgetpendingvalidators) verifies that our transaction was accepted.

Go back to the `Earn` tab, and click `Estimated Rewards`.

Once your validator’s start time has passed, you will see the rewards it may earn, as well as its start time, end time, and the percentage of its validation period that has passed.

That’s it!

## Add a validator with AvalancheJS

We can also add a node to the validator set using [AvalancheJS](../../apis/avalanchejs/README.md).

### Install AvalancheJS

To use AvalancheJS, you can clone the repo:

```zsh
$ git clone https://github.com/ava-labs/avalanchejs.git
```

or add it to an existing project:

```zsh
$ yarn add --dev avalanche
```

For this tutorial we will use [`ts-node`](https://www.npmjs.com/package/ts-node) to run the example scripts directly from an AvalancheJS directory.

### Fuji Workflow

Open your AvalancheJS directory and select the [**`examples/platformvm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/platformvm) folder to view the source code for the examples scripts.

> We will use the [**`buildAddValidatorTx.ts`**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildAddValidatorTx.ts) script to add a validator.

Reviewer 2: Also please link to the avalancheJS API `pchain.buildAddValidatorTx` which is the focus point of this section.

Reviewer 2: @cgcardona what is your suggestion on linking to the API signature?

Reviewer 2: We should add a link to the API signature at https://docs.avax.network/apis/avalanchejs/classes/api_platformvm.platformvmapi#buildaddvalidatortx using the correct relative path.

Author: Roger. I am working on that now

Author: @yulin-dong

is this excerpt correct? if so which one is more accurate?

To learn more about the `buildAddValidatorTx` API, please click [here](../../apis/avalanchejs/classes/api_platformvm.platformvmapi.md#buildaddvalidatortx).

To learn more about the API behind `buildAddValidatorTx`, please click [here](../../apis/avalanchejs/classes/api_platformvm.platformvmapi.md#buildaddvalidatortx).

Reviewer 2: To learn more about the buildAddValidatorTx API, please click [here](https://github.com/ava-labs/apis/avalanchejs/classes/api_platformvm.platformvmapi.md#buildaddvalidatortx).

> We will use the [**`buildAddValidatorTx.ts`**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildAddValidatorTx.ts) script to add a validator.

Reviewer 2: Also please link to the avalancheJS API `pchain.buildAddValidatorTx` which is the focus point of this section.

`rewardAddresses`

When a validator stops validating the Primary Network, they will receive a reward if they are sufficiently responsive and correct while they validated the Primary Network. These tokens are sent to `rewardAddresses`. The original stake will be sent back to the addresses defined in `toAddresses`.

A validator’s stake is never slashed, regardless of their behavior; they will always receive their stake back when they’re done validating.

By default, the example uses the variable `pAddressStrings` to define `toAddresses`, `fromAddresses`, `changeAddress` and `rewardAddresses`:

```js
const pAddressStrings: string[] = pchain.keyChain().getAddressStrings();
```

This retrieves the P-Chain addresses that belong to the `private key` that appears earlier in the example.

If you wish to change these addresses, please follow the steps below:

Locate this line in the file

```js
const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
...
    pAddressStrings,
    pAddressStrings,
    pAddressStrings,
...

>     pAddressStrings,

Reviewer 2: Which one is for which of the following: `toAddresses`, `fromAddresses`, `changeAddress` and `rewardAddresses`. I am totally lost here.

Reviewer 2: Linking to the API signature is critical here so that it is very clear which params is for what.

Author: Definitely. Please refer to the comment above regarding `pchain.buldAddValditor.ts`

...
  )
```

and change the specified `pAddressStrings` to the P-Chain address of choice. Please use array format when passing in each new P-Chain address.

```js
const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
...
    ["<YOUR-PCHAIN-ADDRESS-HERE>"],
    ["<YOUR-PCHAIN-ADDRESS-HERE>"],
    ["<YOUR-PCHAIN-ADDRESS-HERE>"],
...
    ["<YOUR-PCHAIN-ADDRESS-HERE>"],

> ...

Reviewer 2: Same question above, which address is for each of `toAddresses`, `fromAddresses`, `changeAddress` and `rewardAddresses`?

  )
```

### Execute the Add Validator Script

Now that we have made all of the necessary changes to our example script, it's time to add a validator to the Fuji Network.

Run the command:

```zsh
avlanchejs $ ts-node examples/platformvm/buildAddValidatorTx.ts
```

The response has the transaction ID.

```
avlanchejs $ Success! TXID: 2ftDVwmss5eJk8HFsNVi6a3vWK9s3szZFhEeSY2HCS8xDb8Cra
```

We can check the transaction’s status by running the example script: [`getTxStatus.ts`](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/getTxStatus.ts) following the steps below:

1. Ensure that your [network settings](#network-setting) are correct before running the script.

2. Locate this line in the file

```js
const main = async (): Promise<any> => {
  const txID: string = "x1NLb9JaHkKTXvSRReVSsFwQ38mY7bfD1Ky1BPv721VhrpuSE"
  ...
  }
```

and replace it with the _buildAddValidator_ TXID

```js
const main = async (): Promise<any> => {

>  const txID: string = "<YOUR-TXID-HERE>"

Reviewer 2: ```suggestion
 const txID: string = "2ftDVwmss5eJk8HFsNVi6a3vWK9s3szZFhEeSY2HCS8xDb8Cra"
```

That’s it!

## Add a validator with AvalancheJS

We can also add a node to the validator set using [AvalancheJS](../../apis/avalanchejs/README.md).

### Install AvalancheJS

To use AvalancheJS, you can clone the repo:

```zsh
$ git clone https://github.com/ava-labs/avalanchejs.git
```

or add it to an existing project:

```zsh
$ yarn add --dev avalanche
```

For this tutorial we will use [`ts-node`](https://www.npmjs.com/package/ts-node) to run the example scripts directly from an AvalancheJS directory.

### Fuji Workflow

Open your AvalancheJS directory and select the [**`examples/platformvm`**](https://github.com/ava-labs/avalanchejs/tree/master/examples/platformvm) folder to view the source code for the examples scripts.

We will use the [**`buildAddValidatorTx.ts`**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildAddValidatorTx.ts) script to add a validator.

#### Private Key

Locate this line in the file

```js
const privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`;
```

and replace this with a private key that you control.

```js

> const privKey: string = "<YOUR-PRIVATE-KEY-HERE>";

Reviewer: You should add a node about trying to use a Ledger WHENEVER you can.

Author: Definitely!

I was unaware that we can use a ledger to do this. 
Is [this](https://docs.avax.network/apis/avalanche-wallet-sdk/wallet-classes/#ledger-wallet) something I can use to illustrate using a ledger?

Reviewer 2: This will be addressed by https://ava-labs.atlassian.net/browse/DS-208

#### Stake Amount

In order to validate the Fuji Network, one must stake AVAX.

The sample code assigns the variable, `stakeAmount` to be the minimum amount of AVAX that a node must stake to become a validator(1 AVAX).

```ts
const stakeAmount: any = await pchain.getMinStake();
```

To stake a different amount, please replace the code above with a stake amount of 1 AVAX or greater. Below sets a new value of 10,000 AVAX (`10000000000000` GWEI). Value is set in GWEI format where `1,000,000,000` GWEI = 1 AVAX.

Please follow the Example below for steps to change the stake amount:

Locate this line in the file

```js
  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
    ...
    stakeAmount.minValidatorStake,
    ...
  )
```

and change the default value to a new `BN` value, for example:

```js
  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
    ...
    new BN(1e12),
    ...
  )

> ```

Reviewer 2: This is a convoluted way to say that they just need to put a new value for the stake amount as BN(1e12). Please rewrite this section to make it short to the point that they just need to provide a correct amount to the API.

Author: I have read this multiple times and I personally feel that this is short and clear. I have removed some language but am open to more suggestions.

Reviewer 2: This whole section can be reduced to one paragraph/sentence like this:

Set the the proper staking amount when calling pchain.buildAddValidatorTx by replacing `stakeAmount.minValidatorStake` with a number such as `BN(1e12)` which is 10,000 AVAX, in the example code.

That’s it!

## Add a validator with AvalancheJS

We can also add a node to the validator set using [AvalancheJS](https://docs.avax.network/apis/avalanchejs/). To add a node the Fuji Network, we’ll examine the workflow below.

### Getting started

To use AvalancheJS, you can clone the repo:
```zsh
$ git clone https://github.com/ava-labs/avalanchejs.git
```

or add it to an existing project:

```zsh

> $ yarn add --dev avalanche

Reviewer 2: why `--dev` here?

Author: I saw this workflow from a non-production point of view so it seems more accurate.

Open to changing

Reviewer 2: https://classic.yarnpkg.com/lang/en/docs/cli/add/
`Using --dev or -D will install one or more packages in your [devDependencies](https://classic.yarnpkg.com/en/docs/dependency-types#toc-dev-dependencies).`

Are you suggesting developers to just use avalancheJS in devdependencies?

```

For this tutorial we will use `ts-node` to run the example scripts directly from an AvalancheJS directory.

### Fuji Workflow

Open your AvalancheJS directory and select the [**```examples/platformvm```**](https://github.com/ava-labs/avalanchejs/tree/master/examples/platformvm) folder to view the source code for the examples scripts. 

We will use our [**```buildAddValidatorTx.ts```**](https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/buildAddValidatorTx.ts) script to add a validator.

##### Private Key

Locate this line in the file

```js
const privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`;
```

> and replace this with a private key that you control.

Reviewer 3: Maybe link to appropriate example for getting a pirvate key? https://github.com/ava-labs/avalanchejs/blob/master/examples/platformvm/createKeypair.ts

Reviewer 2: Please update based on the comment above.

Author: committed

https://github.com/ava-labs/avalanche-docs/blob/a8a9c63fe3e51b34e1bcfb0b8c2be41c3bcbf13b/docs/nodes/validate/add-a-validator.md?plain=1#L103

#### Stake Amount

In order to validate the FUJI Network, one must stake AVAX.

The sample code assigns the variable, `stakeAmount` to be the minimum amount of AVAX that a node must stake to become a validator(1 AVAX).
```ts
const main = async (): Promise<any> => {
  const stakeAmount: any = await pchain.getMinStake()
  ...
}
```

To stake a different amount, please replace the code above with a stake amount of 1 AVAX or greater. Below sets a new value of 10,000 AVAX (`10000000000000` GWEI). Value is set in GWEI format where `1,000,000,000` GWEI = 1 AVAX. 

Please follow the Example below for steps to change the stake amount:

Locate this line in the file

```js
  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
    ...
    stakeAmount.minValidatorStake,
    ...
  )
```

and change the default value to a new `BN` value, for example:

```js
  const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
    ...

>     new BN(1e12),

Reviewer 2: This is a convoluted way to say that they just need to put a new value for the stake amount as `BN(1e12)`

    ...
  )
```

> #### Start and End Time

Reviewer 2: This is a duplicate section above (Line 200-203)

The sample code uses `const startTime: BN = UnixNow().add(new BN(60 * 1))` and
`const endTime: BN = startTime.add(new BN(26300000))` to compute the Unix time 1 minute and 304 days in the future to use as the values of `startTime` and `endTime`, respectively.

:::tip
You can create your own unix timestamp [here](https://www.unixtimestamp.com/) or by using the `UnixNow()` method
:::

To create your own start times, please follow the steps below:

1. Locate this line in the file

```ts
const startTime: BN = UnixNow().add(new BN(60 * 1))
const endTime: BN = startTime.add(new BN(26300000))
```

2. Change `startTime` and `endTime` to new `BN` values, for example:

```ts
const startTime: BN = new BN(1654656829) // Wed Jun 08 2022 02:53:49 GMT+0000
const endTime: BN = new BN(1662602029) // Thu Sep 08 2022 01:53:49 GMT+0000
```

#### Addresses

`toAddresses`

An array of addresses as [Buffer](https://github.com/feross/buffer) who receive the staked tokens at the end of the staking period.

`fromAddresses`

An array of addresses as a [Buffer](https://github.com/feross/buffer) who own the staking UTXOs.

> `changeAddresses`

Reviewer 3: Are `change`, `to` and `reward` really arrays? On the platform.addValidator() those are all single address, not an array of them.

Author: This was pulled from `avalanchejs/avalanche/dist/apis/platformvm/api.d.ts` 

```js
     * @param toAddresses An array of addresses as {@link https://github.com/feross/buffer|Buffer} who received the staked tokens at the end of the staking period
     * @param fromAddresses An array of addresses as {@link https://github.com/feross/buffer|Buffer} who own the staking UTXOs the fees in AVAX
     * @param changeAddresses An array of addresses as {@link https://github.com/feross/buffer|Buffer} who gets the change leftover from the fee payment
     * @param rewardAddresses The addresses which will receive the rewards from the delegated stake.
```

Also, I tried to pass a single string as an argument [here](https://github.com/ava-labs/avalanchejs/blob/ffaf4d2ef2bbc62a57685d611e3a98115a2d2195/examples/platformvm/buildAddValidatorTx.ts#L41-L50) but the compiler asked for a string[]

Could use clarity on this

#### Customizing Addresses

There are 4 addresses which are needed when calling `pchain.buildAddValidatorTx`. Only 2 of them can be changed: `toAddresses` and `rewardAddresses`. For backward-compatibility reasons, `fromAddresses` and `changeAddresses`
are just placeholders and are ignored.

`toAddresses`

An array of addresses who receive the staked tokens at the end of the staking period.

`rewardAddresses`

When a validator stops validating the Primary Network, they will receive a reward if they are sufficiently responsive and correct while they validated the Primary Network. These tokens are sent to `rewardAddresses`. The original stake will be sent back to the addresses defined in `toAddresses`.

A validator’s stake is never slashed, regardless of their behavior; they will always receive their stake back when they’re done validating.

If you want to customize the `to`, `from`, `change`, and `reward` addresses, please adapt the example below into your project.

Locate this part of the code

```ts

> let privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`;

Reviewer 4: In our example code we don't use semicolons so please remove the semicolons from each of these example snippets.

Reviewer 2: @Julian-dev28 please update

```ts
let privKey: string = `${PrivateKeyPrefix}${DefaultLocalGenesisPrivateKey}`;
pKeychain.importKey(privKey);
privKey = "PrivateKey-FLKnwy3eFr75LHw6mPn8RAsAcWPq4WD8erMWiTK4KsLvkuP2t";
pKeychain.importKey(privKey);
privKey = "PrivateKey-2AUhohZSb86GKXE3mTBNVQGDWYpgR3FGHh7r6gqFoUzXDBNCcA";
pKeychain.importKey(privKey);
priKey = "PrivateKey-JeiGENXTFw5NWzBR18iacf1mL18VosfHf5eURhvmqSiAZKLaC";
pKeychain.importKey(privKey);

const pAddressStrings: string[] = pchain.keyChain().getAddressStrings();
```

This example would create a keychain with 4 addresses:

```ts
[
  "P-custom18jma8ppw3nhx5r4ap8clazz0dps7rv5u9xde7p", // pAddressStrings[0]
  "P-custom186e8ee4qnhnazrnyv2k9u44607ux70syngdz6l", // pAddressStrings[1]
  "P-custom1pdu8d8ss5p329l4vtn23p25cjr57qt5yw2cnsn", // pAddressStrings[2]
  "P-custom152s2upkznxeucwnp68wmg6wmy863qd2sz6ktww", // pAddressStrings[3]
];
```

Now we can pass in each address according to it's slot in the `pAddressStrings` array:

```ts
const unsignedTx: UnsignedTx = await pchain.buildAddValidatorTx(
  utxoSet,
  [pAddressStrings[0], pAddressStrings[1]], // toAddresses, one or more addresses

>   [], // fromAddresses

Reviewer 4: The `fromAddresses` and `changeAddresses` cannot be empty arrays. They must have at least 1 address. I suggest just passing in an array which contains only the first address from `pAddressStrings`&mdash;`[pAddressStrings[0]]`.

Reviewer 2: addressed
